### ISSUE TITLE
Memory allocation failure in MapController init

### ISSUE THREAD

**Issue by nikhiljha** [tangrams/tangram-es#2215]
**TO REPRODUCE THE ISSUE, FOLLOW THESE STEPS:**

- Use an Android phone with GrapheneOS
- Run an app that uses the tangram library, 0.13.0

**RESULT:**

https://github.com/tangrams/tangram-es/blob/c4b024539f07dff1addc2e6424e7a535a1f0c191/platforms/android/tangram/src/main/java/com/mapzen/tangram/NativeMap.java#L10-L15

```
12-02 10:04:08.935 26927 26927 E AndroidRuntime: FATAL EXCEPTION: main
12-02 10:04:08.935 26927 26927 E AndroidRuntime: Process: de.westnordost.streetcomplete, PID: 26927
12-02 10:04:08.935 26927 26927 E AndroidRuntime: java.lang.RuntimeException: Unable to create a native Map object! There may be insufficient memory available.
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.mapzen.tangram.MapController.<init>(MapController.java:163)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.mapzen.tangram.MapView.getMapInstance(MapView.java:276)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.mapzen.tangram.MapView.initMapController(MapView.java:173)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.mapzen.tangram.MapView$1.onLibraryReady(MapView.java:120)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.mapzen.tangram.MapView$1InitTask.onPostExecute(MapView.java:236)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.mapzen.tangram.MapView$1InitTask.onPostExecute(MapView.java:224)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at android.os.AsyncTask.finish(AsyncTask.java:771)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at android.os.AsyncTask.access$900(AsyncTask.java:199)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:106)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:223)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7656)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.android.internal.os.ExecInit.main(ExecInit.java:43)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
12-02 10:04:08.935 26927 26927 E AndroidRuntime:        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:399)
```

**EXPECTED RESULT:**

- The `NativeMap` object does not throw an error.

**ENVIRONMENT:**

 - What operating system and device did you produce this issue on?
    Pixel 4a, Android 11 (GrapheneOS w/ hardened_malloc `RP1A.201105.002.2020.11.27.15`)
 - If you used a released version, what is the version number?
    [`v0.13.0`](https://github.com/westnordost/StreetComplete/blob/76565cbfdb608528f059698d8613159005ef3332/app/build.gradle.kts#L160)

**OTHER:**

Although this works fine in a regular Android install, failing to allocate with `hardened_malloc` could indicate underlying issues with memory management. Something unsafe is happening for this code path to be triggered.

Thanks 👍

**Comment by nikhiljha** [comment]
These functions also seem relevant but nothing really stands out to me as problematic(?)

https://github.com/tangrams/tangram-es/blob/c4b024539f07dff1addc2e6424e7a535a1f0c191/platforms/android/tangram/src/main/cpp/NativeMap.cpp#L36-L39

https://github.com/tangrams/tangram-es/blob/c4b024539f07dff1addc2e6424e7a535a1f0c191/platforms/android/tangram/src/main/cpp/AndroidMap.cpp#L42-L69

**Comment by matteblair** [comment]
This is super interesting! Sorry it's causing a crash though :\

Do you know of any ways to get diagnostic information from `hardened_malloc`? There are lots of places in the code that allocate memory on startup.

According to the project readme (https://github.com/GrapheneOS/hardened_malloc) it's possible to use `hardened_malloc` with a desktop Linux program. I may be able to use it with Tangram's Linux desktop target to track down the issue.

**Comment by westnordost** [comment]
Okay, @tapetis found something that will help find the culprit.

The crash **only** happens on Android 11 and when the app does not run in any kind of backward compatibility mode (i.e., compiled for and targeting Android 11).

Starting in Android 11, heap allocations have a tag set in the top byte of the pointer on devices with kernel support for ARM Top-byte Ignore (TBI). The feature is described in more detail here: https://source.android.com/devices/tech/debug/tagged-pointers

The linked article also includes some helpful information on what may be the problem if one experiences crashes in the application:

> If your app crashed and you were prompted with this link, it could mean one of the following:
>
> -   The application tried to free a pointer that wasn't allocated by the system's heap allocator.
> -   Something in your app modified the top byte of a pointer. The top byte of the pointer can't be modified and your code needs to be changed to fix this issue.
>
> Examples of the top byte pointer being incorrectly used or modified.
> 
> -   Pointers to a particular type have application specific metadata stored in the top 16 address bits.
> -   A pointer cast to double and then back, losing the lower address bits.
> -   Code computing the difference between the addresses of local variables from different stack frames as a way to measure recursion depth.

**Comment by westnordost** [comment]
So this means, it should be reproducible in the demo app if
- (`compileSdkVersion` and) [`targetSdkVersion` is set to 30](https://github.com/tangrams/tangram-es/blob/master/platforms/android/demo/build.gradle#L10)
- it is being tested on a smartphone with Android 11 or later with a kernel that support TBI. So, not an x86 emulator but an ARM-v8 device. According to the crash reports I got so far, at least the following devices support TBI on Android 11: Google Pixel 4, Google Pixel 4 XL, Google Pixel 5, OnePlus 8. Most manufacturers did not supply updates to Android 11 yet, so the list will probably get much longer (soon-ish).

**Comment by nikhiljha** [comment]
To clarify, the bug happens when either using Android 11 with tagged pointers || hardened_malloc, so that's pretty good evidenc

**Comment by matteblair** [comment]
Wow, this is really useful info! (and some interesting tech in Android!)

The first step for fixing this is a reliable reproduction case. I don't have an Android 11 device, but I can try to run it in an ARMv8 emulator. 

If I can't manage to reproduce the Android crash, I can also try looking for memory management issues with tools like Valgrind and hope that the problem is detectable in the Linux desktop build.

I have some suspects in my head now for what part of the code might be causing this. There is at least one 3rd-party library we use that implements an optimization by storing some data in the unused bytes of pointers - it's easy to disable this behavior if it turns out to be the problem. Another possibility is that we are triggering these pointer checks when we pass pointer values from memory allocated in C++ up to Java as a `long` and then back to C++. However, I believe this practice is very common and is suggested in the Android docs themselves, so it seems unlikely.

**Comment by matteblair** [comment]
Unfortunately, it looks like Google has stopped providing ARM emulator images for Android versions newer than 7 (supposedly the performance is so terrible that they would be essentially useless), so reproducing this issue on an emulator doesn't seem possible. Unless I happen to acquire a phone supporting Android 11, I may have to rely on other users to check whether the issue is resolved!

I also looked more closely at the 3rd-party code that I had been suspicious of. Now I no longer believe that this pointer optimization is to blame because it seems to only be enabled on x86_64 architectures. However, it's worth a try! I've built an AAR with this optimization explicitly disabled. If someone is able to test this on a device where Tangram currently crashes, I would greatly appreciate it!
[tangram-0.13.1-release-NoPtrOpt.aar.zip](https://github.com/tangrams/tangram-es/files/5664636/tangram-0.13.1-release-NoPtrOpt.aar.zip)

**Comment by tapetis** [comment]
I have compiled the StreetComplete app with the non-optimized .aar library, but it still crashes with the same exception. When the TBI check is turned off, the only thing that changed with the new version is that the quest pin (or selection ring) is now offset.

![Screenshot_20201209-111626](https://user-images.githubusercontent.com/6892794/101617161-934b3780-3a10-11eb-86f5-c65a09150806.png)

**Comment by westnordost** [comment]
The offset is another issue that has been fixed for a while on tangram-es master but hasn't been released yet. The offset looks more wrong with the fix because SC intentionally sets a wrong offset to counteract the now fixed offset bug that existed until 0.13

On 9 December 2020 11:42:37 CET, Tim <notifications@github.com> wrote:
>I have compiled the StreetComplete app with the non-optimized .aar
>library, but it still crashes with the same exception. When the TBI
>check is turned off, the only thing that changed with the new version
>is that the quest pin (or selection ring) is now offset.
>
>![Screenshot_20201209-111626](https://user-images.githubusercontent.com/6892794/101617161-934b3780-3a10-11eb-86f5-c65a09150806.png)
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/tangrams/tangram-es/issues/2215#issuecomment-741688387

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by matteblair** [comment]
Ok I think I know what's going on now, and it's way dumber than I first thought.

I had been looking for ways that Tangram could be mis-managing memory and triggering the new pointer checks in Android 11. However, the crash reports have not indicated that the application was killed by the system - the application was killed from an uncaught exception _in our own code_! So then I looked closer at the condition that leads to that exception.

The native init code allocates native memory and returns the pointer to it as a Java `long`. There's a check on this returned value to make sure the allocation succeeded. The Java code throws an exception if `nativePtr <= 0`. Clearly `0` is an error - we can't dereference `NULL` later on. But what about `< 0`? This would only happen when the highest bit in the pointer is set. On a normal system this would never happen, but this is _exactly_ what happens when pointer tagging is enabled!! In that case, those pointers aren't actually invalid, but our code _thinks_ they are because of the way pointer tagging is implemented.

The bottom line is: I am pretty sure this crash will be resolved by changing `nativePtr <= 0` to `nativePtr == 0`. I'll make this change and upload a build for testing tonight.

**Comment by westnordost** [comment]
Oh!

Very nice!

**Comment by matteblair** [comment]
The change is here: https://github.com/tangrams/tangram-es/pull/2217 and attached is an AAR with the fix. @tapetis I'd be grateful if you can test this one too!

[tangram-0.13.1-release-PtrChkFix.aar.zip](https://github.com/tangrams/tangram-es/files/5670618/tangram-0.13.1-release-PtrChkFix.aar.zip)

**Comment by tapetis** [comment]
StreetComplete is still crashing with the fix when and only when pointer tagging is enabled, but now the exception has changed.
```
de.westnordost.streetcomplete.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.westnordost.streetcomplete.debug, PID: 814
    java.lang.RuntimeException: Unable to create new data source
        at com.mapzen.tangram.MapController.addDataLayer(MapController.java:674)
        at de.westnordost.streetcomplete.map.tangram.KtMapController.addDataLayer(KtMapController.kt:276)
        at de.westnordost.streetcomplete.map.tangram.KtMapController.addDataLayer$default(KtMapController.kt:275)
        at de.westnordost.streetcomplete.map.QuestsMapFragment.onMapReady(QuestsMapFragment.kt:78)
        at de.westnordost.streetcomplete.map.MapFragment.initMap(MapFragment.kt:175)
        at de.westnordost.streetcomplete.map.MapFragment$initMap$1.invokeSuspend(Unknown Source:11)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by matteblair** [comment]
Nice! This confirms my theory, it means I just missed some spots where I need to change the pointer comparison. I've updated the branch and here's a new AAR: [tangram-0.13.1-release-PtrChkFix2.aar.zip](https://github.com/tangrams/tangram-es/files/5674733/tangram-0.13.1-release-PtrChkFix2.aar.zip)

**Comment by tapetis** [comment]
With the new fix, the app starts properly most of the time. However, sometimes it crashes with a ```SEGV_MAPERR```.

```
A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x111d5d in tid 26598 (GLThread 5844), pid 26541 (tcomplete.debug)
```
```
Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x8 in tid 29119 (GLThread 5989), pid 29062 (tcomplete.debug)
```

While launching the app very often to try to reproduce this crash, I also encountered some Java ```NullPointerException```s.
```
E/TANGRAM: >>> dispose
E/TANGRAM: <<< http requests: 0
E/MapFragment: Error on disposing map
    java.lang.NullPointerException: Attempt to invoke virtual method 'void com.mapzen.tangram.MapData.remove()' on a null object reference
        at com.mapzen.tangram.MapController.dispose(MapController.java:215)
        at com.mapzen.tangram.MapView.disposeMap(MapView.java:262)
        at com.mapzen.tangram.MapView.onDestroy(MapView.java:312)
        at de.westnordost.streetcomplete.map.MapFragment.onDestroy(MapFragment.kt:147)
        at de.westnordost.streetcomplete.map.LocationAwareMapFragment.onDestroy(LocationAwareMapFragment.kt:113)
        at de.westnordost.streetcomplete.map.QuestsMapFragment.onDestroy(QuestsMapFragment.kt:91)
        at androidx.fragment.app.Fragment.performDestroy(Fragment.java:2928)
        at androidx.fragment.app.FragmentStateManager.destroy(FragmentStateManager.java:492)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1308)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1368)
        at androidx.fragment.app.FragmentManager.moveFragmentToExpectedState(FragmentManager.java:1446)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1509)
        at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2637)
        at androidx.fragment.app.FragmentManager.dispatchDestroy(FragmentManager.java:2621)
        at androidx.fragment.app.Fragment.performDestroy(Fragment.java:2923)
        at androidx.fragment.app.FragmentStateManager.destroy(FragmentStateManager.java:492)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1308)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1368)
        at androidx.fragment.app.FragmentManager.moveFragmentToExpectedState(FragmentManager.java:1446)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1509)
        at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2637)
        at androidx.fragment.app.FragmentManager.dispatchDestroy(FragmentManager.java:2621)
        at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:330)
        at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:365)
        at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:242)
        at android.app.Activity.performDestroy(Activity.java:8245)
        at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1344)
        at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5090)
        at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:5135)
        at android.app.servertransaction.DestroyActivityItem.execute(DestroyActivityItem.java:44)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by matteblair** [comment]
Thanks @tapetis! This looks to me like the original issue is resolved by the proposed changes. Of course, I'll also want to track down these other problems you're seeing. For the sake of organization, can you open a new issue?

**Comment by matteblair** [comment]
The `NullPointerException` you've described might be the same as https://github.com/tangrams/tangram-es/issues/2136

**Comment by tapetis** [comment]
Yes, it looks like they are the same.

I have not yet had time to test whether the `SIGSEV` crash also occurs when pointer tagging is disabled. However, accessing an unmapped memory page indicates that it may also be caused by the randomized bits. If this is not the case, I will open another issue tomorrow. Currently, with TBI enabled, the app immediately crashes at approximately every 10th launch.

**Comment by tapetis** [comment]
The `SIGSEGV` crashes occur both with and without pointer tagging enabled. However, they only appear with the PtrChkFix2 version and not with the 0.13.0 release. I have attached the entire tombstones of multiple `SEGV_ACCERR` and `SEGV_MAPERR` errors.

[SEGV_ACCERR-1.txt](https://github.com/tangrams/tangram-es/files/5679538/SEGV_ACCERR-1.txt)
[SEGV_ACCERR-2.txt](https://github.com/tangrams/tangram-es/files/5679539/SEGV_ACCERR-2.txt)
[SEGV_MAPERR-1.txt](https://github.com/tangrams/tangram-es/files/5679540/SEGV_MAPERR-1.txt)
[SEGV_MAPERR-2.txt](https://github.com/tangrams/tangram-es/files/5679541/SEGV_MAPERR-2.txt)
[SEGV_MAPERR-3.txt](https://github.com/tangrams/tangram-es/files/5679537/SEGV_MAPERR-3.txt)

**Comment by matteblair** [comment]
Original issue resolved by https://github.com/tangrams/tangram-es/pull/2217 - the segfault has been moved to another issue https://github.com/tangrams/tangram-es/issues/2218
